File: anghammarad-client-node/src/aws.ts
import * as SNS from "aws-sdk/clients/sns";
import {
  CredentialProviderChain,
  EC2MetadataCredentials,
  EnvironmentCredentials,
  SharedIniFileCredentials,
} from "aws-sdk";

/**
 * Use this to make an SNS client, or provide your own.
 */
export function snsClient(credentialProvider: CredentialProviderChain): SNS {
  return new SNS({ region: "eu-west-1", credentialProvider });
}

export function credentialsProvider(): CredentialProviderChain {
  return new CredentialProviderChain([
    // EC2
    () => new EC2MetadataCredentials(),
    // Lambda
    () => new EnvironmentCredentials(""),
    // local
    () => new SharedIniFileCredentials({ profile: "deployTools" }),
  ]);
}

File: client/src/main/scala/com/gu/anghammarad/AWS.scala
package com.gu.anghammarad


import java.util.concurrent.CompletableFuture

import software.amazon.awssdk.regions.Region
import software.amazon.awssdk.services.sns.SnsAsyncClient
import software.amazon.awssdk.auth.credentials.AwsCredentialsProviderChain
import software.amazon.awssdk.auth.credentials.{EnvironmentVariableCredentialsProvider, ProfileCredentialsProvider, InstanceProfileCredentialsProvider}



import scala.concurrent.{Future, Promise}


object AWS {
  /**
    * Use this to make an SNS client, or provide your own.
    */
  def snsClient(credentialsProvider: AwsCredentialsProviderChain): SnsAsyncClient = {
    SnsAsyncClient.builder()
      .region(Region.EU_WEST_1)
      .credentialsProvider(credentialsProvider)
      .build()
  }

  def credentialsProvider(): AwsCredentialsProviderChain = {
    AwsCredentialsProviderChain.of(
      // EC2
      InstanceProfileCredentialsProvider.create(),
      // Lambda
      EnvironmentVariableCredentialsProvider.create(),
      // local
      ProfileCredentialsProvider.create("deployTools"),
    )
  }

  private[anghammarad] def asScala[T](cf: CompletableFuture[T]): Future[T] = {
    val p = Promise[T]()                                                                                                                                           
    cf.whenCompleteAsync{ (result, ex) =>                                                                                                                          
      if (result == null) p failure ex                                                                                                                             
      else                p success result                                                                                                                         
    }                                                                                                                                                              
    p.future 
  }

}